p2p: protect onions in AttemptToEvictConnection(), add eviction protection test coverage

jonatack opened the pull request:
Now that #19991 and #20210 have been merged, we can determine inbound onion peers using `CNode::m_inbound_onion` and add it to the localhost peers protection in `AttemptToEvictConnection`, which was added in #19670 to address issue #19500.

Update 28 February 2021: I've updated this to follow gmaxwell's suggestion in https://github.com/bitcoin/bitcoin/pull/20197#issuecomment-713865992.

This branch now protects up to 1/4 onion peers (connected via our tor control service), if any, sorted by longest uptime. If any (or all) onion slots remain after that operation, they are then allocated to protect localhost peers, or a minimum of 2 localhost peers in the case that no onion slots remain and 2 or more onion peers were protected, sorted as before by longest uptime.

This patch also adds test coverage for the longest uptime, localhost, and onion peer eviction protection logic to build on the welcome initial unit testing of #20477.

Suggest reviewing the commits that move code with `colorMoved = dimmed-zebra` and `colorMovedWs = allow-indentation-change`.

Closes #11537.

jonatack committed:
Use EraseLastKElements() throughout SelectNodeToEvict()

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

sdaftuar commented:
Concept ACK

practicalswift commented:
Concept ACK

hebasto commented:
Concept ACK.

jonatack commented:
Added `CNode::m_inbound_onion` unit test coverage in #20120.

gmaxwell commented:
Sounds good, although you might want to add a couple slots for actual localhost peers. :)

sipa commented:
> Sounds good, although you might want to add a couple slots for actual localhost peers. :)

Anyone with an old manually-configured hidden service not using the new bind style will have their inbound Tor connections not classified as such, so this may be degradation for them if there is no explicit affordance for localhost peers.

jonatack commented:
Thank you for the feedback @gmaxwell and @sipa, will update soon.

practicalswift commented:
Now that we have unit testing for the eviction logic in `master` (see merged PR #20477): what about adding a small test for this change? :)

jonatack commented:
Rebased, and updated to fall back to localhost if no onion peers are removed. Depending on feedback, this could go further and count onion peers removed to activate localhost if too few onion peers are removed.

> Now that we have unit testing of the eviction logic in `master` (see merged PR #20477): what about adding a small test for this change? :)

@practicalswift I agree and spent time yesterday writing tests after extracting some of the untested parts of `SelectNodeToEvict()` to a separate function for reliable testability of the current code and plan to propose that in a separate PR.

jonatack committed:
Add m_inbound_onion to AttemptToEvictConnection()

and an `m_is_onion` struct member to NodeEvictionCandidate and tests.

We'll use these in the peer eviction logic to protect inbound onion peers
in addition to the existing protection of localhost peers.

jonatack commented:
Rebased and updated after merge of #19972 that created a silent merge conflict.

vasild commented on src/net.cpp:
```diff
@@ -923,16 +929,25 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
+    // Pick out up to 1/4 peers that are connected via our tor onion service, sorted by longest uptime.
+    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareOnionTimeConnected);
+    const size_t local_erase_size = total_protect_size / 2;
+    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_onion; }), vEvictionCandidates.end());
```
What about extending `EraseLastKElements()` with one more argument "predicate" and remove elements only if it returns `true` on them (with a default that assumes `true` for any element)?

Then this code would look like:

```cpp
EraseLastKElements(
    vEvictionCandidates,
    CompareOnionTimeConnected,
    local_erase_size,
    [](NodeEvictionCandidate const &n) { return n.m_is_onion; });
```

maybe out of the scope of this PR.

vasild replied:
Something like this may work (untested, feel free to ignore):

<details>
<summary>[patch] EraseLastKElements() with predicate</summary>

```diff
diff --git i/src/net.cpp w/src/net.cpp
index ddf33c38a..64900b8b5 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -41,12 +41,13 @@
 // with Ubuntu 16.04 LTS and Debian 8 libminiupnpc-dev packages.
 static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed");
 #endif
 
 #include <algorithm>
 #include <cstdint>
+#include <functional> /* std::function */
 #include <unordered_map>
 
 #include <math.h>
 
 /** Maximum number of block-relay-only anchor connections */
 static constexpr size_t MAX_BLOCK_RELAY_ONLY_ANCHORS = 2;
@@ -893,38 +894,46 @@ static bool CompareNodeBlockRelayOnlyTime(const NodeEvictionCandidate &a, const
     if (a.fRelayTxes != b.fRelayTxes) return a.fRelayTxes;
     if (a.nLastBlockTime != b.nLastBlockTime) return a.nLastBlockTime < b.nLastBlockTime;
     if (a.fRelevantServices != b.fRelevantServices) return b.fRelevantServices;
     return a.nTimeConnected > b.nTimeConnected;
 }
 
-//! Sort an array by the specified comparator, then erase the last K elements.
-template<typename T, typename Comparator>
-static void EraseLastKElements(std::vector<T> &elements, Comparator comparator, size_t k)
+//! Sort an array by the specified comparator, then erase each one of the last K elements if
+//! `pred(element)` returns true.
+template <typename T, typename Comparator>
+static void EraseLastKElements(
+    std::vector<T>& elements,
+    Comparator comparator,
+    size_t k,
+    std::function<bool(const T&)> pred = [](const T&) { return true; })
 {
     std::sort(elements.begin(), elements.end(), comparator);
     size_t eraseSize = std::min(k, elements.size());
-    elements.erase(elements.end() - eraseSize, elements.end());
+    elements.erase(std::remove_if(elements.end() - eraseSize, elements.end(), pred),
+                   elements.end());
 }
 
 [[nodiscard]] Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates)
 {
     // Protect connections with certain characteristics
 
+    std::function<bool(const NodeEvictionCandidate&)> pred;
+
     // Deterministically select 4 peers to protect by netgroup.
     // An attacker cannot predict which netgroups will be protected
     EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4);
     // Protect the 8 nodes with the lowest minimum ping time.
     // An attacker cannot manipulate this metric without physically moving nodes closer to the target.
     EraseLastKElements(vEvictionCandidates, ReverseCompareNodeMinPingTime, 8);
     // Protect 4 nodes that most recently sent us novel transactions accepted into our mempool.
     // An attacker cannot manipulate this metric without performing useful work.
     EraseLastKElements(vEvictionCandidates, CompareNodeTXTime, 4);
     // Protect up to 8 non-tx-relay peers that have sent us novel blocks.
-    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareNodeBlockRelayOnlyTime);
-    size_t erase_size = std::min(size_t(8), vEvictionCandidates.size());
-    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return !n.fRelayTxes && n.fRelevantServices; }), vEvictionCandidates.end());
+    const size_t erase_size = std::min(size_t(8), vEvictionCandidates.size());
+    pred = [](const NodeEvictionCandidate& n) { return !n.fRelayTxes && n.fRelevantServices; };
+    EraseLastKElements(vEvictionCandidates, CompareNodeBlockRelayOnlyTime, erase_size, pred);
 
     // Protect 4 nodes that most recently sent us novel blocks.
     // An attacker cannot manipulate this metric without performing useful work.
     EraseLastKElements(vEvictionCandidates, CompareNodeBlockTime, 4);
 
     // Protect the half of the remaining nodes which have been connected the longest.
@@ -933,22 +942,22 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
     // they're not longest-uptime overall. This helps protect tor peers, which
     // tend to be otherwise disadvantaged under our eviction criteria.
     const size_t initial_size = vEvictionCandidates.size();
     size_t total_protect_size = initial_size / 2;
 
     // Pick out up to 1/4 peers that are connected via our tor onion service, sorted by longest uptime.
-    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareOnionTimeConnected);
     const size_t local_erase_size = total_protect_size / 2;
-    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_onion; }), vEvictionCandidates.end());
+    pred = [](const NodeEvictionCandidate& n) { return n.m_is_onion; };
+    EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, local_erase_size, pred);
 
     // If no onion peers were removed, extend the same protection to localhost peers, as manually
     // configured hidden services not using -bind will not be detected as inbound onion connections.
     if (initial_size == vEvictionCandidates.size()) {
         // Pick out up to 1/4 peers that are localhost, sorted by longest uptime.
-        std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareLocalHostTimeConnected);
-        vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_local; }), vEvictionCandidates.end());
+        pred = [](const NodeEvictionCandidate& n) { return n.m_is_local; };
+        EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, local_erase_size, pred);
     }
 
     // Calculate how many we removed, and update our total number of peers that
     // we want to protect based on uptime accordingly.
     total_protect_size -= initial_size - vEvictionCandidates.size();
     EraseLastKElements(vEvictionCandidates, ReverseCompareNodeTimeConnected, total_protect_size);
@@ -956,14 +965,15 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
     if (vEvictionCandidates.empty()) return nullopt;
 
     // If any remaining peers are preferred for eviction consider only them.
     // This happens after the other preferences since if a peer is really the best by other criteria (esp relaying blocks)
     //  then we probably don't want to evict it no matter what.
     if (std::any_of(vEvictionCandidates.begin(),vEvictionCandidates.end(),[](NodeEvictionCandidate const &n){return n.prefer_evict;})) {
-        vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.begin(),vEvictionCandidates.end(),
-                                  [](NodeEvictionCandidate const &n){return !n.prefer_evict;}),vEvictionCandidates.end());
+        pred = [](const NodeEvictionCandidate& n) { return !n.prefer_evict; };
+        EraseLastKElements(vEvictionCandidates, ReverseCompareNodeTimeConnected,
+                           vEvictionCandidates.size(), pred);
     }
 
     // Identify the network group with the most connections and youngest member.
     // (vEvictionCandidates is already sorted by reverse connect time)
     uint64_t naMostConnections;
     unsigned int nMostConnections = 0;
```
</details>

jonatack replied:
Thanks! I started with the quick patch below on January 2, will compare with your ideas above.

<details><summary>patch</summary><p>

```diff
-//! Sort an array by the specified comparator, then erase the last K elements.
+//! Sort an array by the specified comparator, then erase the last K elements where predicate is true.
 template<typename T, typename Comparator>
-static void EraseLastKElements(std::vector<T> &elements, Comparator comparator, size_t k)
+static void EraseLastKElements(std::vector<T> &elements, Comparator comparator, size_t k, std::function<bool(const NodeEvictionCandidate&)> predicate = [](NodeEvictionCandidate const &n) { return true ;})
 {
     std::sort(elements.begin(), elements.end(), comparator);
-    size_t eraseSize = std::min(k, elements.size());
-    elements.erase(elements.end() - eraseSize, elements.end());
+    size_t erase_size = std::min(k, elements.size());
+    elements.erase(std::remove_if(elements.end() - erase_size, elements.end(), predicate), elements.end());
 }
 
 [[nodiscard]] Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates)
@@ -934,16 +934,15 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
     size_t total_protect_size = initial_size / 2;
 
     // Pick out up to 1/4 peers that are connected via our tor onion service, sorted by longest uptime.
-    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareOnionTimeConnected);
     const size_t local_erase_size = total_protect_size / 2;
-    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_onion; }), vEvictionCandidates.end());
+    EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, local_erase_size, [](NodeEvictionCandidate const &n) { return n.m_is_onion; });
+;
 
     // If no onion peers were removed, extend the same protection to localhost peers, as manually
-    // configured hidden services not using -bind will not be detected as inbound onion connections.
+    // configured hidden services not using -bind=addr:port=onion will not be detected as inbound onion connections.
     if (initial_size == vEvictionCandidates.size()) {
         // Pick out up to 1/4 peers that are localhost, sorted by longest uptime.
-        std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareLocalHostTimeConnected);
-        vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_local; }), vEvictionCandidates.end());
+        EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, local_erase_size, [](NodeEvictionCandidate const &n) { return n.m_is_local; });
     }
```
</p></details>

vasild commented on src/net.cpp:
```diff
@@ -923,16 +929,25 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
+
+    // If no onion peers were removed, extend the same protection to localhost peers, as manually
+    // configured hidden services not using -bind will not be detected as inbound onion connections.
+    if (initial_size == vEvictionCandidates.size()) {
```
Just to note - this assumes that if there are recognizable onion connections (via `-bind=addr:port=onion`) then there are **not** non-recognizable ones (statically/manually configured tor daemon pointing to our standard p2p port, e.g. 8333).

This assumption may not be true as there could be a node that is configured both via `-bind=addr:port=onion` (connections coming to 8334 are recognized as onion) and also a static tor config redirecting to 8333.

I think the code is ok as it is because the above would be rare and maybe not worth the code complexity to tweak the eviction policy for it.

vasild reviewed (approved):
ACK c16ed372bdc71795fbebc5dbf519a1d748a136dc

The code looks ok, after some staring at it. Just some minor notes below.

jnewbery commented on src/test/fuzz/node_eviction.cpp:
```diff
@@ -21,17 +21,18 @@ FUZZ_TARGET(node_eviction)
-            fuzzed_data_provider.ConsumeIntegral<uint64_t>(),
-            fuzzed_data_provider.ConsumeBool(),
-            fuzzed_data_provider.ConsumeBool(),
+            /* id */ fuzzed_data_provider.ConsumeIntegral<NodeId>(),
+            /* nTimeConnected */ fuzzed_data_provider.ConsumeIntegral<int64_t>(),
```
Designated initializers are only standardized in c++20: https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers

vasild replied:
Already used here?

https://github.com/bitcoin/bitcoin/blob/11d3b5833671db3b884171086489e853738e7ddd/src/test/fuzz/util.h#L272-L273

jnewbery replied:
Interesting. Are the fuzzers compiled with --std=c++2a?

vasild replied:
No. My guess is that fuzz tests are not compiled with the old compiler that chokes on this syntax and that most compilers support it even in `-std=c++17` mode.

jonatack commented on src/net.h:
```diff
@@ -1247,6 +1247,7 @@ struct NodeEvictionCandidate
     uint64_t nKeyedNetGroup;
     bool prefer_evict;
     bool m_is_local;
+    bool m_is_onion;
```
Sure, I think the salient change in that commit is adding `m_inbound_onion` to `AttemptToEvictConnection` but happy to append "and add m_is_onion to NodeEvictionCandidate" to the commit message body.

jonatack replied:
Done.

jonatack commented on src/net.cpp:
```diff
@@ -923,16 +929,25 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
+    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_onion; }), vEvictionCandidates.end());
+
+    // If no onion peers were removed, extend the same protection to localhost peers, as manually
+    // configured hidden services not using -bind will not be detected as inbound onion connections.
```
Thanks, replaced `-bind` with `-bind=<addr>[:<port>]=onion` like the -bind help.

sdaftuar commented:
> Sounds good, although you might want to add a couple slots for actual localhost peers. :)

I interpreted this comment from @gmaxwell as suggesting that we should protect some actual localhost peers, even if we are also protecting some onion peers -- looks like this patch right now would only try to protect some localhost peers if there are no onion peers that we're protecting.  Any reason not to protect some of each?  I think that having more categories/metrics that we use to distinguish inbounds will make it harder for an adversary to take over all our inbound slots.

jonatack committed:
Move peer eviction tests to a separate test file

out of net_tests, because the eviction tests:

- are a different domain of test coverage, with different dependencies

- run more slowly than the net tests

- will be growing in size, in this PR branch and in the future, as eviction
  test coverage is improved

jonatack committed:
Protect onion+localhost peers in ProtectEvictionCandidatesByRatio()

Now that we have a reliable way to detect inbound onion peers, this commit
updates our existing eviction protection of 1/4 localhost peers to instead
protect up to 1/4 onion peers (connected via our tor control service), sorted by
longest uptime. Any remaining slots of the 1/4 are then allocated to protect
localhost peers, or 2 localhost peers if no slots remain and 2 or more onion
peers are protected, sorted by longest uptime.

The goal is to avoid penalizing onion peers, due to their higher min ping times
relative to IPv4 and IPv6 peers, and improve our diversity of peer connections.

Thank you to Gregory Maxwell, Suhas Daftuar, Vasil Dimov and Pieter Wuille
for valuable review feedback that shaped the direction.

jonatack committed:
Add unit tests for ProtectEvictionCandidatesByRatio()

Thank you to Vasil Dimov (vasild) for the suggestion to use std::unordered_set
rather than std::vector for the IsProtected() peer id arguments.

jonatack committed:
Extract ProtectEvictionCandidatesByRatio from SelectNodeToEvict

to allow deterministic unit testing of the ratio-based peer eviction protection
logic, which protects peers having longer connection times and those connected
via higher-latency networks.

Add documentation.

jonatack committed:
Add unit test coverage for our onion peer eviction protection

jonatack commented:
Thank you @vasild  and @sdaftuar. Updated to potentially protect up to 1/4 onion peers, independently of the existing 1/4 localhost protection, for an unchanged 1/2 of the remaining peers overall.

Open to suggestions regarding the ratios. Some ideas, given that we'll have I2P support once #20685 is merged (a draft of I2P eviction protection is at #21261):

```
Onion        1/6    1/4     1/6     1/6
Localhost    1/6    1/10    1/6     1/12
I2P          1/6    1/10    1/8     1/12
CJDNS                               1/12
```


Added test coverage for the existing protection of half the remaining peers sorted by longest uptime, of 1/4 localhost peers, and finally of 1/4 onion peers.

Suggest reviewing the move-only commits with `colorMoved = dimmed-zebra` and `colorMovedWs = allow-indentation-change`

jonatack commented:
Improved the last test commit, adding combined onion + localhost tests.

jonatack commented:
Found and fixed a bug in `IsProtected()` in the tests. Improved, simplified and sped them up.

vasild reviewed (commented):
Approach ACK f1c02b3ef9fa4bbd82144eebbeaf8c9c9df238c5

About how many to protect - I do not feel strongly about it, but I think we should protect 1/4 (localhost or tor), not 1/4 localhost and 1/4 tor. Because localhost likely means a tor peer that somehow is not recognized as such. So, a tor peer is (surely recognized as tor OR unrecognized one, coming from localhost (likely)) and the original code aims to protect 1/4 tor peers and my understanding is that this PR's purpose is not to possibly extend that to 1/2 (1/4 + 1/4).

Just to be explicit what I mean:
<details>
<summary>[patch] protect 1/4 cumulatively tor+localhost</summary>

```diff
diff --git i/src/net.cpp w/src/net.cpp
index e0e0ba9d8..96a1259e7 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -827,23 +827,24 @@ static bool ReverseCompareNodeMinPingTime(const NodeEvictionCandidate &a, const
 
 static bool ReverseCompareNodeTimeConnected(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
     return a.nTimeConnected > b.nTimeConnected;
 }
 
-static bool CompareLocalHostTimeConnected(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
-{
-    if (a.m_is_local != b.m_is_local) return b.m_is_local;
-    return a.nTimeConnected > b.nTimeConnected;
-}
-
-static bool CompareOnionTimeConnected(const NodeEvictionCandidate& a, const NodeEvictionCandidate& b)
-{
-    if (a.m_is_onion != b.m_is_onion) return b.m_is_onion;
-    return a.nTimeConnected > b.nTimeConnected;
-}
+/**
+ * Compare eviction candidates in such a way, that (localhost or onion) ones are at
+ * the end. Within them put the ones that have been connected longer (have smaller
+ * `nTimeConnected`) at the end.
+ */
+static bool CompareThisIsHard(const NodeEvictionCandidate& a, const NodeEvictionCandidate& b)
+{
+    const bool a_is_special = a.m_is_local || a.m_is_onion;
+    const bool b_is_special = b.m_is_local || b.m_is_onion;
+    if (a_is_special != b_is_special) return b_is_special;
+    return a.nTimeConnected > b.nTimeConnected;
+}
 
 static bool CompareNetGroupKeyed(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b) {
     return a.nKeyedNetGroup < b.nKeyedNetGroup;
 }
 
 static bool CompareNodeBlockTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
@@ -891,21 +892,18 @@ void ProtectEvictionCandidates(std::vector<NodeEvictionCandidate>& vEvictionCand
     // they're not longest-uptime overall. This helps protect tor peers, which
     // tend to be otherwise disadvantaged under our eviction criteria.
     const size_t initial_size = vEvictionCandidates.size();
     size_t total_protect_size = initial_size / 2;
     const size_t local_erase_size = total_protect_size / 2;
 
-    // Pick out up to 1/4 peers connected via our onion service, sorted by longest uptime.
-    EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, local_erase_size,
-                       [](const NodeEvictionCandidate& n) { return n.m_is_onion; });
-
-    // Pick out up to 1/4 peers that are localhost, sorted by longest uptime,
-    // as manually configured hidden services not using `-bind=addr[:port]=onion`
-    // will not be detected as inbound onion connections.
-    EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, local_erase_size,
-                       [](const NodeEvictionCandidate& n) { return n.m_is_local; });
+    // Pick out up to 1/4 localhost peers (likely they are tor ones that do not benefit from
+    // `-bind=addr[:port]=onion`) or peers connected via our onion service, sorted by longest uptime.
+    EraseLastKElements(vEvictionCandidates,
+                       CompareThisIsHard,
+                       local_erase_size,
+                       [](const NodeEvictionCandidate& n) { return n.m_is_local || n.m_is_onion; });
 
     // Calculate how many we removed, and update our total number of peers that
     // we want to protect based on uptime accordingly.
     total_protect_size -= initial_size - vEvictionCandidates.size();
     EraseLastKElements(vEvictionCandidates, ReverseCompareNodeTimeConnected, total_protect_size);
 }
```

</details>

Again, I do not have a strong opinion on this. Curious to see what other reviewers think.

jonatack committed:
Use std::unordered_set instead of std::vector in IsEvicted()

An unordered set can tell if an element is present in ~O(1) time (constant on
average, worst case linear to the size of the container), which speeds up and
simplifies the lookup in IsEvicted().

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack commented on src/test/net_peer_eviction_tests.cpp:
```diff
@@ -0,0 +1,304 @@
+    }
+    Shuffle(candidates.begin(), candidates.end(), random_context);
+
+    const size_t expected{candidates.size() / 2}; // expect half the candidates will be protected
+    ProtectEvictionCandidates(candidates);
+    const size_t actual{candidates.size()};
+    BOOST_CHECK(actual == expected || actual == expected + 1);
```
That's better indeed. Done.

jonatack commented on src/net.cpp:
```diff
@@ -889,39 +924,24 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
-        vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.begin(),vEvictionCandidates.end(),
-                                  [](NodeEvictionCandidate const &n){return !n.prefer_evict;}),vEvictionCandidates.end());
+    if (std::any_of(vEvictionCandidates.begin(), vEvictionCandidates.end(), [](const NodeEvictionCandidate& n) { return n.prefer_evict; })) {
+        EraseLastKElements(vEvictionCandidates, ReverseCompareNodeTimeConnected, vEvictionCandidates.size(),
+                           [](const NodeEvictionCandidate& n) { return !n.prefer_evict; });
```
Good idea. Decided for now to not touch this line instead.

jonatack commented on src/test/net_peer_eviction_tests.cpp:
```diff
@@ -0,0 +1,304 @@
+    // independently of other characteristics.
+    BOOST_CHECK(IsProtected(
+        num_peers, [](NodeEvictionCandidate& c) {
+            c.m_is_onion = (c.id == 3 || c.id == 8 || c.id == 9) ? true : false;
```
Done, thank you.

jonatack commented on src/test/net_peer_eviction_tests.cpp:
```diff
@@ -0,0 +1,304 @@
+    BOOST_CHECK(actual == expected || actual == expected + 1);
+
+    size_t unprotected_as_expected{0};
+    for (const NodeEvictionCandidate& candidate : candidates) {
+        if (std::find(protected_peer_ids.begin(), protected_peer_ids.end(), candidate.id) != protected_peer_ids.end()) {
+            // this peer should have been removed from the eviction candidates
+            BOOST_TEST_MESSAGE(strprintf("candidate id: %d", candidate.id));
+            return false;
+        }
+        if (std::find(unprotected_peer_ids.begin(), unprotected_peer_ids.end(), candidate.id) != unprotected_peer_ids.end()) {
+            // this peer remains in the eviction candidates, as expected
+            ++unprotected_as_expected;
+        }
+    }
```
Nice, thank you! And no need for the `> 0`. Done for both `IsProtected` and `IsEvicted`.

jonatack commented:
Thank you @vasild for your excellent review! I've taken your inline code feedback.

WRT the question of "how many to protect and how to do it," I think it may be best to continue picking out the onion peers and the localhost peers separately in order to strive for more peer diversity. The idea is that if we do somehow have both onion and localhost peers, it seems worthwhile to ensure we protect some of each, even if one has worse latency than the other.

Thus, at the cost of a wee bit more complexity, I've updated this to follow @gmaxwell's suggestion more closely.

This branch now protects up to 1/4 onion peers (connected via our tor control service), if any, sorted by longest uptime. If any tor slots remain after that operation, they are then allocated to protect localhost peers, or a minimum of 2 localhost peers in the case that no tor slots remain and 2 or more onion peers were protected, sorted as before by longest uptime.

Therefore, of the half of the remaining peers which have been connected the longest that we currently protect, this patch reserves a maximum of (half + 2) of them for onion and localhost peers. E.g., if we have 12 remaining peers, we'll protect half (6), and up to half of those (3) could be reserved for onions plus up to 2 for localhost.

jonatack commented:
Improved the unit tests.

jonatack commented:
Rebased. @naumenkogs, @ariard, @jnewbery, @dongcarl, as reviewers of #19670 you might be interested in this. @practicalswift, this updates your unit tests and adds missing ones as you had requested. @wtogami, this might address the issues you discussed with me recently.

ariard commented on src/net.h:
```diff
@@ -1287,8 +1288,11 @@ struct NodeEvictionCandidate
     uint64_t nKeyedNetGroup;
     bool prefer_evict;
     bool m_is_local;
+    bool m_is_onion;
```
Maybe called it `m_is_onion_service` to dissociate from the manually configured hidden services you mentioned above?

jonatack replied:
Thanks @ariard. The latter ones would be `m_is_local`. I think I prefer to go with `m_is_onion` for `CNode::m_inbound_onion` and following the naming style of `m_is_local`.

ariard reviewed (commented):
Concept ACK.

What type of localhost peers coming from automated-connections-but-manually-configured do we have right now ? I2P? Otherwise if you want to protect localhost peers from manual connections maybe better just to redirect them on a `whitebind=noban@addr` ?

vasild commented:
https://github.com/bitcoin/bitcoin/issues/11537 is related to this PR.

vasild reviewed (approved):
ACK ea9f19ee66ff66e1f5c384c19def994ec3037e13

vasild commented:
@ariard, I2P peers are not localhost ones. They can easily be distinguished because their address's `CNetAddr::IsI2P()` is true. The complication with Tor is because incoming tor connections come from an IP address (not from a Tor address), usually `127.0.0.1`.

I am not sure what you mean by "automated-connections-but-manually-configured" or "protect localhost peers from manual connections".

jonatack commented:
Rebased due to conflicts from the merge of #21415 --  ` git range-diff a9d1b40 ea9f19e c18fd59`

vasild reviewed (approved):
ACK c18fd59329a855c12ba370474a3ec39032611f2b

sipa reviewed (commented):
utACK c18fd59329a855c12ba370474a3ec39032611f2b. I didn't review the last (unit test) commit in detail.

jonatack commented on src/net.cpp:
```diff
@@ -879,6 +879,26 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
     elements.erase(elements.end() - eraseSize, elements.end());
 }
 
+void ProtectEvictionCandidates(std::vector<NodeEvictionCandidate>& vEvictionCandidates)
```
Good idea. Done, renamed the function to `ProtectEvictionCandidatesByRatio()` (open to suggestions) and added Doxygen documentation.

jonatack replied:
Also updated the commit names/descriptions.

jonatack commented:
Added documentation and renamed a function per https://github.com/bitcoin/bitcoin/pull/20197#discussion_r597365954.

<details><summary><code>git diff c18fd59 0cca08a</code></summary><p>

```diff
diff --git a/src/net.cpp b/src/net.cpp
index be93426377..66d81b115f 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -887,7 +887,7 @@ static void EraseLastKElements(
     elements.erase(std::remove_if(elements.end() - eraseSize, elements.end(), predicate), elements.end());
 }
 
-void ProtectEvictionCandidates(std::vector<NodeEvictionCandidate>& vEvictionCandidates)
+void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvictionCandidates)
 {
     // Protect the half of the remaining nodes which have been connected the longest.
     // This replicates the non-eviction implicit behavior, and precludes attacks that start later.
@@ -945,7 +945,9 @@ void ProtectEvictionCandidates(std::vector<NodeEvictionCandidate>& vEvictionCand
     // An attacker cannot manipulate this metric without performing useful work.
     EraseLastKElements(vEvictionCandidates, CompareNodeBlockTime, 4);
 
-    ProtectEvictionCandidates(vEvictionCandidates);
+    // Protect some of the remaining eviction candidates by ratios of desirable
+    // or disadvantaged characteristics.
+    ProtectEvictionCandidatesByRatio(vEvictionCandidates);
 
     if (vEvictionCandidates.empty()) return std::nullopt;
 
diff --git a/src/net.h b/src/net.h
index 449e0c54b7..eb7fa079ab 100644
--- a/src/net.h
+++ b/src/net.h
@@ -1285,8 +1285,36 @@ struct NodeEvictionCandidate
     bool m_is_onion;
 };
 
-void ProtectEvictionCandidates(std::vector<NodeEvictionCandidate>& vEvictionCandidates);
-
+/**
+ * Select an inbound peer to evict after filtering out (protecting) peers having
+ * distinct, difficult-to-forge characteristics. The protection logic picks out
+ * fixed numbers of desirable peers per various criteria, followed by (mostly)
+ * ratios of desirable or disadvantaged peers. If any eviction candidates
+ * remain, the selection logic chooses a peer to evict.
+ */
 [[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates);
 
+/** Protect desirable or disadvantaged inbound peers from eviction by ratio.
+ *
+ * This function protects half of the peers which have been connected the
+ * longest, to replicate the non-eviction implicit behavior and preclude attacks
+ * that start later.
+ *
+ * Half of these protected spots (1/4 of the total) are reserved for onion peers
+ * connected via our tor control service, if any, sorted by longest uptime, even
+ * if they're not longest uptime overall. Any remaining slots of the 1/4 are
+ * then allocated to protect localhost peers, if any (or up to 2 localhost peers
+ * if no slots remain and 2 or more onion peers were protected), sorted by
+ * longest uptime, as manually configured hidden services not using
+ * `-bind=addr[:port]=onion` will not be detected as inbound onion connections.
+ *
+ * This helps protect onion peers, which tend to be otherwise disadvantaged
+ * under our eviction criteria for their higher min ping times relative to IPv4
+ * and IPv6 peers, and favorise the diversity of peer connections.
+ *
+ * This function was extracted from SelectNodeToEvict() to be able to test the
+ * ratio-based protection logic deterministically.
+ */
+void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvictionCandidates);
+
 #endif // BITCOIN_NET_H
diff --git a/src/test/net_peer_eviction_tests.cpp b/src/test/net_peer_eviction_tests.cpp
index 7b88a3e9b6..31d391bf7d 100644
--- a/src/test/net_peer_eviction_tests.cpp
+++ b/src/test/net_peer_eviction_tests.cpp
@@ -43,9 +43,9 @@ std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_c
 }
 
 // Create `num_peers` random nodes, apply setup function `candidate_setup_fn`,
-// apply protection logic, and then return true if all of `protected_peer_ids`
-// and none of `unprotected_peer_ids` are protected from eviction, i.e. removed
-// from the eviction candidates.
+// call ProtectEvictionCandidatesByRatio() to apply protection logic, and then
+// return true if all of `protected_peer_ids` and none of `unprotected_peer_ids`
+// are protected from eviction, i.e. removed from the eviction candidates.
 bool IsProtected(int num_peers,
                  std::function<void(NodeEvictionCandidate&)> candidate_setup_fn,
                  const std::unordered_set<NodeId>& protected_peer_ids,
@@ -60,7 +60,7 @@ bool IsProtected(int num_peers,
 
     const size_t size{candidates.size()};
     const size_t expected{size - size / 2}; // Expect half the candidates will be protected.
-    ProtectEvictionCandidates(candidates);
+    ProtectEvictionCandidatesByRatio(candidates);
     BOOST_CHECK_EQUAL(candidates.size(), expected);
 ```
</p></details>

vasild reviewed (approved):
ACK 0cca08a8ee33b4e05ff586ae4fd914f5ea860cea

prayank23 commented:
Interesting PR. Concept ACK. Compiled successfully on Ubuntu. Tests passed.

laanwj commented:
Code review ACK 0cca08a8ee33b4e05ff586ae4fd914f5ea860cea
I did not do anything in terms of manual testing, but it looks like the change is well-covered by tests.

mzumsande commented on src/net.cpp:
```diff
@@ -893,30 +937,17 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
-    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareNodeBlockRelayOnlyTime);
-    size_t erase_size = std::min(size_t(8), vEvictionCandidates.size());
-    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return !n.fRelayTxes && n.fRelevantServices; }), vEvictionCandidates.end());
+    const size_t erase_size = std::min(size_t(8), vEvictionCandidates.size());
```
Since `EraseLastKElements` determines the minimum again, this line seems unnecessary (could just pass `size_t(8)`).

jonatack replied:
Good eye! Looks like this was already the case before this PR but can sneak in the update in #21261.

jonatack replied:
Thanks again @mzumsande. Addressed in #21261.

mzumsande reviewed (commented):
Post-Merge Comment (reviewed but didn't finish in time yesterday evening):

I understand the motivation, but just noting that the fixed  `localhost_min_protect_size` in the ratio-based `ProtectEvictionCandidatesByRatio` makes things unintuitive at a lower `initial_size`:

Both cases are nicely covered by the unit tests, but it doesn't make a lot of sense to me that at 6 total peers, we'd protect 1  onion, 0 localhost and 2 by generic uptime, whereas at 8 peers, we'd protect 2 onion, 2 localhost and 0 by generic uptime (raising the percentage of potentially protected tor peers to 50% instead of the targeted 25%, so that no slots are left for protecting generic uptime peers).

jonatack commented:
Thanks for having a look! Here we have an initial improvement and unit testing in place, and in the next step in #21261 we'll re-evaluate the algorithm and ratios between onion, localhost and I2P peers (considering that other privacy networks may be added soon as well). Please do weigh in there with the ratios you would suggest.

jonatack commented:
@mzumsande I've updated the inbound eviction protection proposal in #21261 to be generalizable to multiple networks (with I2P and CJDNS on the way) and fully ratio-based, which should address your feedback. Will be bringing it out of draft shortly.
